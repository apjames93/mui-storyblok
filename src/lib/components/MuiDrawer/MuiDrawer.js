import React from 'react';
import PropTypes from 'prop-types';
import { Drawer } from '@material-ui/core';
import StoryBlok from '../../utils/Storyblok';
import {
  validComponents,
  muiStringProp,
} from '../../utils/customProps';
import { renderComponents } from '../../utils/customComponents';
import MuiList from '../MuiList/MuiList';
import MuiListDropdown from '../MuiListDropdown/MuiListDropdown';

const MuiDrawer = ({
  anchor,
  rootClass,
  elevation,
  variant,
  content,
  onClose,
  open,
}) => {
  const components = {
    MuiList,
    MuiListDropdown,
  };

  let width = '100%';
  if (anchor === 'left' || anchor === 'right') width = '32vw';
  const styles = StoryBlok.arrayToMuiStyles(rootClass, { minWidth: width });

  // eslint-disable-next-line no-multi-assign
  const handleToggleDrawer = (event) => {
    if (event.type === 'keydown' && (event.key === 'Tab' || event.key === 'Shift')) {
      return;
    }
    onClose(event);
  };

  return (
    <Drawer
      className={styles.root}
      anchor={anchor}
      classes={{ paper: styles.root }}
      elevation={+elevation}
      variant={variant}
      open={open}
      onClose={e => handleToggleDrawer(e)}
    >
      {content.map((component, key) => renderComponents(components, component, key)) }
    </Drawer>
  );
};

export default MuiDrawer;

MuiDrawer.propTypes = {
  /**
   * storyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),

  /**
   * mui props: 'bottom', 'left', 'right', 'top'
   * Side from which the drawer will appear.
   */
  anchor(props, propName, componentName) {
    const validProps = ['bottom', 'left', 'right', 'top'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * number
   * The elevation of the drawer.
   * storyblok type is a number
   */
  elevation: PropTypes.string,
  /**
   * mui props: 'permanent', 'persistent', 'temporary'
   * The variant to use.
   */
  variant(props, propName, componentName) {
    const validProps = ['permanent', 'persistent', 'temporary'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * components:
   * MuiList, MuiListDropdown
   */
  content(props, propName, componentName) {
    const components = ['MuiList', 'MuiListDropdown'];
    return validComponents(props, propName, componentName, components);
  },


  open: PropTypes.bool,
  onClose: PropTypes.func,
};

MuiDrawer.defaultProps = {
  anchor: 'left',
  elevation: 16,
  variant: 'permanent',
  rootClass: [],
  content: [],
  open: false,
  onClose: undefined,
};
