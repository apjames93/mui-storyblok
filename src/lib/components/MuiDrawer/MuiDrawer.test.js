import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import MuiDrawer from './MuiDrawer';

function setup(open = false) {
  const props = {
    open,
    onClose: jest.fn(),
    elevation: 16,
    variant: 'temporary',
    content: [
      {
        component: 'MuiList',
        content: [
          {
            component: 'MuiListItem',
            listItemText: [
              {
                primary: 'apples',
                component: 'MuiListItemText',
              },
            ],
            listItemSecondaryAction: [
              {
                component: 'MuiListItemSecondaryAction',
                content: [
                  {
                    component: 'MuiIconButtonRedirect',
                    redirectRoute: 'home',
                    iconButton: [
                      {
                        component: 'MuiIconButton',
                        icon: [
                          {
                            component: 'MuiIcon',
                            iconName: 'home',
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  };
  const comp = mount(<MuiDrawer {...props} />);
  return { comp, props };
}

describe('<MuiDrawer />', () => {
  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(<MuiDrawer {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should handleClose and toggle drawer', () => {
    const { comp, props } = setup(true);
    const closeModal = comp.find('ForwardRef(Modal)').first().prop('onClose');
    act(() => {
      closeModal({ type: 'keydown', key: 'Tab' });
    });
    expect(comp.find('WithStyles(ForwardRef(Drawer))').first().props().open).toEqual(true);
    expect(props.onClose).not.toHaveBeenCalled();
    act(() => {
      closeModal({ type: 'keydown', key: 'Shift' });
    });
    expect(comp.find('WithStyles(ForwardRef(Drawer))').first().props().open).toEqual(true);
    expect(props.onClose).not.toHaveBeenCalled();

    act(() => {
      closeModal({ target: {} });
    });
    expect(props.onClose).toBeCalled();
  });

  it('should render MuiDrawer', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });
});
