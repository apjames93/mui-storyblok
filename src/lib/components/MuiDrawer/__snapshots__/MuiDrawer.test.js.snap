// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiDrawer /> snapshot 1`] = `null`;
